File: lib/components/component.go
package components

import (
	"context"

	mqtt "github.com/eclipse/paho.mqtt.golang"
	"github.com/gregdel/homed/lib/config"
	"github.com/prometheus/client_golang/prometheus"
	"go.uber.org/zap"
)

// Type reprensents a component type
type Type string

// Types
var (
	// Common components
	TypeGenericSensor Type = "generic_sensor"
	TypeBinarySensor  Type = "binary_sensor"
	TypePowerMeter    Type = "power_meter"
	TypeCounter       Type = "counter"
	TypeSwitch        Type = "switch"
	TypeBinaryLight   Type = "binary_light"
	TypeWifiSignal    Type = "wifi_signal"
	TypeDeviceStatus  Type = "device_status"
	TypeRollerShutter Type = "roller_shutter"

	// Other components
	TypeHomedTemperature    Type = "homed_temperature"
	TypeHomedHumidity       Type = "homed_humidity"
	TypeZigbeeTRV           Type = "zigbee_trv"
	TypeBoiler              Type = "boiler"
	TypeZigbeeClimateSensor Type = "zigbee_climate_sensor"
)

// Component represents a component
type Component interface {
	// Type represents the component type
	Type() Type

	// FriendlyName
	FriendlyName() string

	// Device
	Device() *Device
	SetDevice(d *Device)

	// Config
	Config() *config.Component
	SetConfig(*config.Component)

	// Logger
	LoggerWithFields(*zap.Logger) *zap.Logger

	// Update is called every time a new mqtt payload is received on the state
	// topic
	Update([]byte) error

	// PostUpdate is a function called after the Update function
	PostUpdate() error

	// Collector returns the prometheus collectors to register for the component
	Collectors(labels prometheus.Labels) []prometheus.Collector

	// ReadOnly tells if the component is readonly
	ReadOnly() bool
	// Internal() tells if the component is internal
	Internal() bool

	// WriteCommand writes a command to a mqtt topic (for external components)
	WriteCommand(data []byte) error
	// ExecCommand executes a command (for internal components)
	ExecCommand(data []byte) error
	// PublishToStateTopic publishes raw data to the state topic
	PublishToStateTopic(data []byte) error

	SetMQTTClient(mqtt.Client)
	MQTTClient() mqtt.Client

	// Run runs a goroutine for a component
	Run(context.Context, *zap.Logger, *Components) error

	// ID
	ID() string
	SetID(id string)

	// Subscribe
	Subscribe(string, chan Event)
	Notify()
}

File: frontend/src/js/components/Components/Component.js
import React from "react";
import PropTypes from "prop-types";

import { Switch } from "./Switch";
import { PowerMeter } from "./PowerMeter";
import { BinaryLight } from "./BinaryLight";
import { BinaryFan } from "./BinaryFan";
import { BinaryTRV } from "./BinaryTRV";

import { Boiler } from "./Boiler";
import { DeviceStatus } from "./DeviceStatus";
import { RollerShutter } from "./RollerShutter";
import { GenericSensor } from "./GenericSensor";
import { BinarySensor } from "./BinarySensor";
import { HomedTemperature } from "./HomedTemperature";
import { TemperatureSwitch } from "./TemperatureSwitch";
import { WifiSignal } from "./WifiSignal";
import { ZigbeeClimateSensor } from "./ZigbeeClimateSensor";
import { VirtualSwitch } from "./VirtualSwitch";
import { HeaderCard } from "./common/HeaderCard";

export const Component = ({ id, type, noCard = false }) => {
  var typedComponent;
  switch (type) {
    case "generic_sensor":
      typedComponent = <GenericSensor id={id} />;
      break;
    case "binary_sensor":
      typedComponent = <BinarySensor id={id} />;
      break;
    case "device_status":
      typedComponent = <DeviceStatus id={id} />;
      break;
    case "wifi_signal":
      typedComponent = <WifiSignal id={id} />;
      break;
    case "boiler":
      typedComponent = <Boiler id={id} />;
      break;
    case "binary_light":
      typedComponent = <BinaryLight id={id} />;
      break;
    case "binary_fan":
      typedComponent = <BinaryFan id={id} />;
      break;
    case "binary_trv":
      typedComponent = <BinaryTRV id={id} />;
      break;
    case "switch":
      typedComponent = <Switch id={id} />;
      break;
    case "roller_shutter":
      typedComponent = <RollerShutter id={id} />;
      break;
    case "homed_temperature":
      typedComponent = <HomedTemperature id={id} />;
      break;
    case "homed_temperature_switch":
      typedComponent = <TemperatureSwitch id={id} />;
      break;
    case "zigbee_climate_sensor":
      typedComponent = <ZigbeeClimateSensor id={id} />;
      break;
    case "power_meter":
      typedComponent = <PowerMeter id={id} />;
      break;
    case "virtual_switch":
      typedComponent = <VirtualSwitch id={id} />;
      break;
    default:
      typedComponent = <>Unhandled {type}</>;
      break;
  }

  if (noCard) {
    return typedComponent;
  }

  return <HeaderCard id={id}>{typedComponent}</HeaderCard>;
};
Component.propTypes = {
  id: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  noCard: PropTypes.bool,
};
